commands_test.go

```diff
@@ -6305,6 +6314,116 @@ var _ = Describe("Commands", func() {
+			Expect(x.Err()).NotTo(HaveOccurred())
+			Expect(x.Int()).To(Equal(3))
+		})
+
+		It("Calls a read-only function", func() {
+			lib1Code = fmt.Sprintf(lib1.Code, lib1.Name, lib1.Functions[0].Name,
+				lib1.Functions[0].Description, lib1.Functions[0].Flags[0], lib1.Functions[0].Flags[1])
+
+			err := client.FunctionLoad(ctx, lib1Code).Err()
+			Expect(err).NotTo(HaveOccurred())
+
+			// This function doesn't have a "no-writes" flag
+			x := client.FCallRo(ctx, lib1.Functions[0].Name, []string{"my_hash"}, "a", 1, "b", 2)
+
+			Expect(x.Err()).To(HaveOccurred())
+
+			lib2Code = fmt.Sprintf(lib2.Code, lib2.Name, lib2.Functions[0].Name, lib2.Functions[1].Name,
+				lib2.Functions[1].Description, lib2.Functions[1].Flags[0])
+
+			// This function has a "no-writes" flag
+			err = client.FunctionLoad(ctx, lib2Code).Err()
+			Expect(err).NotTo(HaveOccurred())
+
+			x = client.FCallRo(ctx, lib2.Functions[1].Name, []string{})
+
+			Expect(x.Err()).NotTo(HaveOccurred())
+			Expect(x.Text()).To(Equal("Function 2"))
+		})
+
+		It("Shows function stats", func() {
+			// The 5M iterations in the f1 function should block for around 3-4 seconds when executed
+			lib := redis.Library{
+				Name:   "mylib1",
+				Engine: "LUA",
+				Functions: []redis.Function{
+					{
+						Name:        "lib1_func1",
+						Description: "This is the func-1 of lib 1",
+						Flags:       []string{"allow-stale"},
+					},
+				},
+				Code: `#!lua name=%s
+					
+					local function f1(keys, args)
+                     local t1 = redis.call('TIME')[1]
+                     for i = 5000000, 1,-1 do 
+                     redis.call('SET', 'a', i)
+                     end
+                     local t2 = redis.call('TIME')[1]
+               
+                     redis.log(redis.LOG_NOTICE, t2-t1)
+                     return t2 - t1
+					end
+
+					redis.register_function{
+						function_name='%s',
+						description ='%s',
+						callback=f1,
+						flags={'%s'}
+					}`,
+			}
+			libCode := fmt.Sprintf(lib.Code, lib.Name, lib.Functions[0].Name,
+				lib.Functions[0].Description, lib.Functions[0].Flags[0])
+			err := client.FunctionLoad(ctx, libCode).Err()
+
+			Expect(err).NotTo(HaveOccurred())
+
+			r, err := client.FunctionStats(ctx).Result()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(r.Engines)).To(Equal(1))
+			Expect(r.RunningScript.IsEmpty()).To(BeTrue())
+
+			started := make(chan bool)
+			done := make(chan bool)
+			go func() {
+				defer GinkgoRecover()
+
+				started <- true
+				callResult := client.FCall(ctx, lib.Functions[0].Name, []string{})
+				Expect(callResult.Err()).NotTo(HaveOccurred())
+				done <- true
+			}()
+
+			<-started
+			time.Sleep(1 * time.Second)
+			r, err = client.FunctionStats(ctx).Result()
+
+			select {
+			case <-done:
+				Expect(err).NotTo(HaveOccurred())
+				Expect(len(r.Engines)).To(Equal(1))
+				Expect(r.RunningScript.Name).To(Equal(lib.Functions[0].Name))
```

Comment:
@vmihailenco @monkey92t I need some help making this test pass. I can't seem to get any result for `running_script` from `function stats`.

Reply:
> 

We need a way to block lua.

Reply:
That's the thing, blocking commands are not allowed inside of functions. I was hoping that the 5M iterations of `SET` would be enough. On my machine, they take around 3-4 seconds.